docs/upcoming_changes/9330.change.rst

```diff
@@ -0,0 +1,23 @@
+Refactor and remove legacy APIs/testing internals.
+""""""""""""""""""""""""""""""""""""""""""""""""""
+
+A number of internally used functions have been removed to aid with general
+maintenance by reducing the number of ways in which it is possible to invoke
+compilation, specifically:
+
+* ``numba.core.compiler.compile_isolated`` is removed.
+* ``numba.tests.support.TestCase::run_nullary_func`` is removed.
+* ``numba.tests.support.CompilationCache`` is removed.
+
+Additionally, the concept of "nested context" is removed from
+``numba.core.registry.CPUTarget`` along with the implementation details.
+Maintainers of target extensions (those using the
+API in ``numba.core.target_extension`` to extend Numba support to
+custom/synthetic hardware) should note that the same can be deleted from
+target extension implementations of ``numba.core.descriptor.TargetDescriptor``
+if it present. i.e. the ``nested_context`` method and associated implementation
```

Comment:
actually, do you even need "if it present" here. the sentence is already long

Reply:
I wanted the `if` in there to make it so that users are unconcerned if their implementation doesn't have one as it's not a requirement. For example the CUDA target does not have a `nested_context`:
https://github.com/numba/numba/blob/688c5141e96eb81dc365a864419c1141b2949333/numba/cuda/descriptor.py#L10-L30